diff --git a/CMakeLists.txt b/CMakeLists.txt
index 78c0345..813e3d1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -124,7 +124,7 @@ else (LUA)
     if (USE_LUAJIT)
       # We only link the libs on Windows, so find_package fully succeeding
       # is only required on Windows
-      if (WIN32)
+      if (WIN32 OR CYGWIN)
         find_package(LuaJIT REQUIRED)
         link_directories(${LUAJIT_LIBRARIES})
       else()
@@ -137,7 +137,7 @@ else (LUA)
     else (USE_LUAJIT)
       # We only link the libs on Windows, so find_package fully succeeding
       # is only required on Windows
-      if (WIN32)
+      if (WIN32 OR CYGWIN)
         find_package(Lua REQUIRED)
       else()
         find_package(Lua)
@@ -207,18 +207,13 @@ if (BUILD_MODULE)
 endif (BUILD_MODULE)
 if (BUILD_STATIC_LIBS)
   add_library(libluv_a STATIC src/luv.c)
-  set_target_properties(libluv_a PROPERTIES PREFIX "")
+  set_target_properties(libluv_a PROPERTIES OUTPUT_NAME luv)
   list(APPEND ACTIVE_TARGETS "libluv_a")
 endif (BUILD_STATIC_LIBS)
 if (BUILD_SHARED_LIBS)
   add_library(libluv SHARED src/luv.c)
-  set_target_properties(libluv PROPERTIES PREFIX "")
   set_target_properties(libluv
-    PROPERTIES VERSION ${LUV_VERSION} SOVERSION ${LUV_VERSION_MAJOR})
-  # Prevent lib prefix in shared library name
-  if(WIN32)
-    set_target_properties(libluv PROPERTIES IMPORT_PREFIX "")
-  endif(WIN32)
+    PROPERTIES VERSION ${LUV_VERSION} SOVERSION ${LUV_VERSION_MAJOR} OUTPUT_NAME luv)
   if(APPLE)
     set_target_properties(libluv
       PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
@@ -249,7 +244,7 @@ if(WIN32)
 endif()
 
 foreach(TARGET_NAME ${ACTIVE_TARGETS})
-  if(WIN32)
+  if(WIN32 OR CYGWIN)
     if (LUA)
       target_link_libraries(${TARGET_NAME} ${LIBUV_LIBRARIES} ${LUA_LIBRARIES})
     else (LUA)
