diff --git a/configure.ac b/configure.ac
index 028d5559..94f99385 100644
--- a/configure.ac
+++ b/configure.ac
@@ -585,6 +585,10 @@ case "$host_os" in
 		AC_MSG_RESULT(cygwin)
 		PLATFORM=cygwin
 		;;
+	*msys*)
+		AC_MSG_RESULT(msys)
+		PLATFORM=msys
+		;;
 	*)
 		AC_MSG_RESULT(unknown)
 		PLATFORM=unknown
diff --git a/options-table.c b/options-table.c
index e3362ba8..a30825a0 100644
--- a/options-table.c
+++ b/options-table.c
@@ -490,7 +490,7 @@ const struct options_table_entry options_table[] = {
 	  .type = OPTIONS_TABLE_STRING,
 	  .scope = OPTIONS_TABLE_SESSION,
 	  .default_str = "DISPLAY SSH_ASKPASS SSH_AUTH_SOCK SSH_AGENT_PID "
-	                 "SSH_CONNECTION WINDOWID XAUTHORITY"
+	                 "SSH_CONNECTION WINDOWID XAUTHORITY MSYSTEM"
 
 	},
 
diff --git a/server-client.c b/server-client.c
index 21cccc85..3d079d0c 100644
--- a/server-client.c
+++ b/server-client.c
@@ -1262,7 +1262,7 @@ server_client_dispatch_identify(struct client *c, struct imsg *imsg)
 		return;
 	c->flags |= CLIENT_IDENTIFIED;
 
-#ifdef __CYGWIN__
+#if defined(__CYGWIN__) || defined(__MSYS__)
 	c->fd = open(c->ttyname, O_RDWR|O_NOCTTY);
 #endif
 
diff --git a/tmux.c b/tmux.c
index 53b984d7..7682eb0b 100644
--- a/tmux.c
+++ b/tmux.c
@@ -150,7 +150,7 @@ make_label(const char *label)
 		errno = ENOTDIR;
 		goto fail;
 	}
-	if (sb.st_uid != uid || (sb.st_mode & S_IRWXO) != 0) {
+	if (sb.st_uid != uid) {
 		errno = EACCES;
 		goto fail;
 	}
